kong/clustering.lua

```diff
@@ -0,0 +1,409 @@
+  if err then
+    ngx_log(ngx_ERR, "unable to ping control plane node: ", err)
+    -- return and let the main thread handle the error
+    return
+  end
+
+  ngx_log(ngx_DEBUG, "sent PING packet to control plane")
+end
+
+
+local function communicate(premature, conf)
+  if premature then
+    -- worker wants to exit
+    return
+  end
+
+  -- TODO: pick one random CP
+  local address = conf.cluster_control_plane
+
+  local c = assert(ws_client:new(WS_OPTS))
+  local uri = "wss://" .. address .. "/v1/outlet?node_id=" ..
+              kong.node.get_id() .. "&node_hostname=" .. utils.get_hostname()
+  local res, err = c:connect(uri, { ssl_verify = true,
+                                    client_cert = CERT,
+                                    client_priv_key = CERT_KEY,
+                                    server_name = "kong_clustering",
+                                  }
+                            )
+  if not res then
+    local delay = 9 + math.random()
+
+    ngx_log(ngx_ERR, "connection to control plane broken: ", err,
+            " retrying after ", delay , " seconds")
+    assert(ngx.timer.at(delay, communicate, conf))
+    return
+  end
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      send_ping(c)
+      ngx_sleep(PING_INTERVAL)
+    end
+  end)
+
+  while true do
+    local data, typ, err = c:recv_frame()
+    if err then
+      ngx.log(ngx.ERR, "error while receiving frame from control plane: ", err)
+      c:close()
+
+      local delay = 9 + math.random()
+      assert(ngx.timer.at(delay, communicate, conf))
+      return
+    end
+
+    if typ == "binary" then
+      local msg = assert(cjson_decode(data))
+
+      if msg.type == "reconfigure" then
+        local config_table = assert(msg.config_table)
+
+        local res, err = update_config(config_table, true)
+        if not res then
+          ngx_log(ngx_ERR, "unable to update running config: ", err)
+        end
+
+        send_ping(c)
+
+      end
+    elseif typ == "pong" then
+      ngx_log(ngx_DEBUG, "received PONG frame from control plane")
+    end
+  end
+end
+
+
+function _M.handle_cp_websocket()
+  -- use mutual TLS authentication
+  local cert = ngx_var.ssl_client_raw_cert
+
+  if not cert then
+    ngx_log(ngx_ERR, "Data Plane failed to present client certificate " ..
+                     "during handshake")
+    return ngx_exit(444)
+  end
+
+  cert = assert(openssl_x509.new(cert, "PEM"))
+  local digest = cert:digest("sha256")
+
+  if CERT_DIGEST and digest ~= CERT_DIGEST then
+    ngx_log(ngx_ERR, "Data Plane presented incorrect client certificate " ..
+                     "during handshake, expected digest: " .. CERT_DIGEST ..
+                     " got: " .. digest)
+    return ngx_exit(444)
+  end
+
+  local node_id = ngx_var.arg_node_id
+  if not node_id then
+    ngx_exit(400)
+  end
+
+  local node_hostname = ngx_var.arg_node_hostname
+  local node_ip = ngx_var.remote_addr
```

Comment:
I don't think there exists a `real_ip` configuration for this. E.g. a dataplane may be behind proxy or lb.

Reply:
This won't work, as LB won't be able to terminate the mTLS connection between DP and CP in order to maintain E2E security, therefore no `XFF` header can not be added.